ql/src/java/org/apache/hadoop/hive/ql/plan/TopNKeyDesc.java

```diff
@@ -252,7 +252,8 @@ public TopNKeyDescExplainVectorization getTopNKeyVectorization() {
   public TopNKeyDesc combine(CommonKeyPrefix commonKeyPrefix) {
     return new TopNKeyDesc(topN, commonKeyPrefix.getMappedOrder(),
             commonKeyPrefix.getMappedNullOrder(), commonKeyPrefix.getMappedColumns(),
-            commonKeyPrefix.getMappedColumns().subList(0, partitionKeyColumns.size()),
+            commonKeyPrefix.getMappedColumns()
+                    .subList(0, Math.min(partitionKeyColumns.size(), commonKeyPrefix.getMappedColumns().size())),
```

Comment:
Yes, this prevents the `ArrayIndexOutOfBoundsException`. However the rest of the patch is also useful since executing TNKs which all keys are partition keys is a waste of CPU time: such instances are always forward operators. 

TNK operators maintain `TopNKeyFilters` for each partition key values. A row is processed by the TNK op. like:
1. extract the partition key values from the row
2. lookup the corresponding `TopNKeyFilter`
3. extract the topN key values from the row and pass it to the `TopNKeyFilter`
4. The `TopNKeyFilter` stores the top n keys and decides whether the row can be forwarded.

Reply:
This new change guarantees that the following condition `keys >= partitionKeys` is always true no matter the input (since the partition keys are a sublist of keys). Maybe, I am missing something but I cannot see how we can have `keys < partitionKeys` at this stage.

If what I am saying is correct then the inequality in the new bail out condition is misleading.

More generally does it make sense to create a new TNK operator when `partitionKeyColumns.size() >= commonKeyPrefix.size()`? If not then maybe we should add appropriate assertion instead of using `Math.min`.